File: agent/unity_tools/refresh.py
"""
THE COMPILER: unity_refresh
"I need to compile my code."
Consumes: refresh_assets

Note: This tool uses LangGraph's interrupt() to pause execution while Unity compiles.
The interrupt is handled by the agent harness which communicates with Unity via WebSocket.
"""
import json
from typing import Optional, List
from pydantic import BaseModel, Field
from langchain_core.tools import StructuredTool
from langgraph.types import interrupt


class RefreshSchema(BaseModel):
    watched_scripts: Optional[List[str]] = Field(
        None,
        description="List of specific script names (e.g. ['PlayerController']) to verify existence of after compilation."
    )
    type_limit: int = Field(
        20,
        description="Limit the number of returned available types to save tokens."
    )


def _unity_refresh(watched_scripts: Optional[List[str]] = None, type_limit: int = 20) -> str:
    """
    Trigger Unity Asset Database refresh and Script Compilation.
    This is "The Compiler".

    CRITICAL: You MUST use this tool after creating or editing C# scripts (.cs files).
    Unity cannot add a component until the script is compiled.

    Behavior:
    1. Pauses agent execution while Unity compiles (handled by orchestrator).
    2. Returns 'compilationErrors' if syntax errors exist.
    3. Confirms if 'watched_scripts' are now valid components.
    """

    # ---------------------------------------------------------
    # BUILD THE REQUEST (will be sent to Unity by harness)
    # ---------------------------------------------------------
    params = {"typeLimit": type_limit}
    if watched_scripts:
        params["watchedScripts"] = watched_scripts

    compile_request = {
        "action": "refresh_assets",
        "params": params
    }

    # ---------------------------------------------------------
    # INTERRUPT: Pause agent, let harness call Unity via WebSocket
    # ---------------------------------------------------------
    # The harness catches this, sends command to Unity via WebSocket,
    # waits for compilation to complete, then resumes with Unity's response
    print(f"DEBUG: Pausing for compilation... request={compile_request}")
    result = interrupt(compile_request)  # <-- Returns Unity's response after resume
    print(f"DEBUG: Resumed from compilation. result={result}")

    # ---------------------------------------------------------
    # HANDLE TIMEOUT - Tell agent to check logs
    # ---------------------------------------------------------
    if not result.get("success") and "timeout" in result.get("error", "").lower():
        return json.dumps({
            "status": "TIMEOUT",
            "message": "Compilation timed out after 40 seconds. This usually means Unity encountered an issue during domain reload.",
            "action_required": "Use unity_query with query_type='editor_log' to check Unity's console for errors or warnings.",
            "common_causes": [
                "Syntax error preventing compilation",
                "Unity Editor dialog popup blocking (API Updater, etc.)",
                "Script with infinite loop in static constructor",
                "Missing assembly reference"
            ]
        }, indent=2)

    # ---------------------------------------------------------
    # SMART RESPONSE PARSING (your existing logic)
    # ---------------------------------------------------------
    body = result.get("body", {})
    success = result.get("success", False)

    # Case 1: Compilation Failed
    if not success or body.get("compilationErrors"):
        errors = body.get("compilationErrors", [])
        return json.dumps({
            "status": "COMPILATION_FAILED",
            "message": "Unity failed to compile the scripts. You must fix these errors:",
            "errors": errors
        }, indent=2)

    # Case 2: Success
    response = {
        "status": "SUCCESS",
        "message": "Assets refreshed and scripts compiled.",
    }

    # Did we find the scripts the agent cared about?
    if watched_scripts and "watchedScriptsStatus" in body:
        response["verification"] = body["watchedScriptsStatus"]

        # Helper text for the LLM
        missing = [name for name, found in body["watchedScriptsStatus"].items() if not found]
        if missing:
            response["warning"] = f"Compilation passed, but these types are still missing: {missing}. Did you get the class name right inside the file?"
        else:
            response["next_step"] = "You can now use unity_component(action='add') with these scripts."

    return json.dumps(response, indent=2)


# Create the tool using StructuredTool
# Note: This tool is synchronous because interrupt() is synchronous
# The async behavior is handled by the LangGraph runtime
unity_refresh = StructuredTool.from_function(
    func=_unity_refresh,
    name="unity_refresh",
    description="""Trigger Unity Asset Database refresh and Script Compilation. This is "The Compiler".

CRITICAL: You MUST use this tool after creating or editing C# scripts (.cs files).
Unity cannot add a component until the script is compiled.

Behavior:
1. Pauses agent execution while Unity compiles (handled by orchestrator).
2. Returns 'compilationErrors' if syntax errors exist.
3. Confirms if 'watched_scripts' are now valid components.""",
    args_schema=RefreshSchema,
)

File: src/agent/unity-tools/refresh.ts
/**
 * THE COMPILER: unity_refresh
 * "I need to compile my code."
 * Consumes: refresh_assets
 *
 * Note: This tool uses LangGraph's interrupt() to pause execution while Unity compiles.
 * The interrupt is handled by the agent harness which communicates with Unity via WebSocket.
 */

import { z } from 'zod';
import { DynamicStructuredTool } from '@langchain/core/tools';
import { interrupt } from '@langchain/langgraph';

/**
 * Zod schema for unity_refresh tool input
 */
export const RefreshSchema = z.object({
    watched_scripts: z.array(z.string()).optional()
        .describe("List of specific script names (e.g. ['PlayerController']) to verify existence of after compilation."),

    type_limit: z.number().int().default(20)
        .describe('Limit the number of returned available types to save tokens.')
});

/** Type inferred from the Zod schema */
export type RefreshInput = z.infer<typeof RefreshSchema>;

/** Response structure for successful compilation */
interface CompilationSuccessResponse {
    status: 'SUCCESS';
    message: string;
    verification?: Record<string, boolean>;
    warning?: string;
    next_step?: string;
}

/** Response structure for failed compilation */
interface CompilationFailedResponse {
    status: 'COMPILATION_FAILED';
    message: string;
    errors: string[];
}

/** Response structure for timeout */
interface CompilationTimeoutResponse {
    status: 'TIMEOUT';
    message: string;
    action_required: string;
    common_causes: string[];
}

/** Unity response body structure */
interface UnityCompileBody {
    compilationErrors?: string[];
    watchedScriptsStatus?: Record<string, boolean>;
}

/** Unity response structure */
interface UnityResponse {
    success: boolean;
    body?: UnityCompileBody;
    error?: string;
}

/**
 * Trigger Unity Asset Database refresh and Script Compilation.
 * This is "The Compiler".
 *
 * CRITICAL: You MUST use this tool after creating or editing C# scripts (.cs files).
 * Unity cannot add a component until the script is compiled.
 */
async function unityRefreshImpl(input: RefreshInput, _config?: any): Promise<string> {
    const { watched_scripts, type_limit = 20 } = input;

    // ---------------------------------------------------------
    // BUILD THE REQUEST (will be sent to Unity by harness)
    // ---------------------------------------------------------
    const params: Record<string, unknown> = { typeLimit: type_limit };
    if (watched_scripts) {
        params.watchedScripts = watched_scripts;
    }

    const compileRequest = {
        action: 'refresh_assets',
        params
    };

    // ---------------------------------------------------------
    // INTERRUPT: Pause agent, let harness call Unity via WebSocket
    // ---------------------------------------------------------
    // The harness catches this, sends command to Unity via WebSocket,
    // waits for compilation to complete, then resumes with Unity's response
    console.log(`DEBUG: Pausing for compilation... request=${JSON.stringify(compileRequest)}`);
    const result = interrupt(compileRequest) as UnityResponse;
    console.log(`DEBUG: Resumed from compilation. result=${JSON.stringify(result)}`);

    // ---------------------------------------------------------
    // HANDLE TIMEOUT - Tell agent to check logs
    // ---------------------------------------------------------
    if (!result.success && result.error?.toLowerCase().includes('timeout')) {
        const timeoutResponse: CompilationTimeoutResponse = {
            status: 'TIMEOUT',
            message: 'Compilation timed out after 40 seconds. This usually means Unity encountered an issue during domain reload.',
            action_required: "Use unity_query with action: 'get_logs' to check Unity's console for errors or warnings.",
            common_causes: [
                'Syntax error preventing compilation',
                'Unity Editor dialog popup blocking (API Updater, etc.)',
                'Script with infinite loop in static constructor',
                'Missing assembly reference'
            ]
        };
        return JSON.stringify(timeoutResponse, null, 2);
    }

    // ---------------------------------------------------------
    // SMART RESPONSE PARSING
    // ---------------------------------------------------------
    const body = result.body ?? {};
    const success = result.success;

    // Case 1: Compilation Failed
    if (!success || body.compilationErrors) {
        const errors = body.compilationErrors ?? [];
        const failedResponse: CompilationFailedResponse = {
            status: 'COMPILATION_FAILED',
            message: 'Unity failed to compile the scripts. You must fix these errors:',
            errors
        };
        return JSON.stringify(failedResponse, null, 2);
    }

    // Case 2: Success
    const response: CompilationSuccessResponse = {
        status: 'SUCCESS',
        message: 'Assets refreshed and scripts compiled.'
    };

    // Did we find the scripts the agent cared about?
    if (watched_scripts && body.watchedScriptsStatus) {
        response.verification = body.watchedScriptsStatus;

        // Helper text for the LLM
        const missing = Object.entries(body.watchedScriptsStatus)
            .filter(([_, found]) => !found)
            .map(([name]) => name);

        if (missing.length > 0) {
            response.warning = `Compilation passed, but these types are still missing: ${JSON.stringify(missing)}. Did you get the class name right inside the file?`;
        } else {
            response.next_step = "You can now use unity_component({ action: 'add' }) with these scripts.";
        }
    }

    return JSON.stringify(response, null, 2);
}

/**
 * The Compiler - unity_refresh tool
 * Trigger Unity Asset Database refresh and Script Compilation.
 *
 * Note: This tool is synchronous because interrupt() is synchronous.
 * The async behavior is handled by the LangGraph runtime.
 */
export const unityRefresh = new DynamicStructuredTool({
    name: 'unity_refresh',
    description: `Trigger Unity Asset Database refresh and Script Compilation. This is "The Compiler".

CRITICAL: You MUST use this tool after creating or editing C# scripts (.cs files).
Unity cannot add a component until the script is compiled.

Behavior:
1. Pauses agent execution while Unity compiles (handled by orchestrator).
2. Returns 'compilationErrors' if syntax errors exist.
3. Confirms if 'watched_scripts' are now valid components.`,
    schema: RefreshSchema,
    func: unityRefreshImpl
});
